File: kernel/epsilon/epsilon_core/riemannian_optimizer.py
# Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
# SPDX-License-Identifier: Epsilon-Hollow

"""
Epsilon-Hollow — Riemannian Gradient Flow Optimizer
====================================================
Manifold-aware gradient descent on curved spaces.

Standard SGD operates in flat Euclidean space ℝ^n. When weights live on
a manifold (orthogonality constraints, unit-norm embeddings, subspace tracking),
Euclidean gradients can violate manifold constraints. This optimizer
performs updates along geodesics using the exponential map.

Mathematical Foundation:
    1. Riemannian Gradient:
       ∇_M f(x) = Π_{T_x M}(∇f(x))
       where Π projects the Euclidean gradient onto the tangent space T_x M.

    2. Exponential Map (geodesic step):
       x_{t+1} = Exp_x(-η · ∇_M f(x))
       On Stiefel manifold St(n, p) = {X ∈ ℝ^{n×p} : X^T X = I_p}:
           Exp_X(Z) = (X + Z)(I + Z^T Z)^{-1/2}  [Cayley retraction, fast]

    3. Parallel Transport:
       Momentum must be transported along the manifold:
       Γ_{x→y}(v) = (I - ½ δδ^T / ‖δ‖²) v   where δ = y - x
       [Vector transport by projection, Absil et al. 2008]

    4. Grassmann Manifold Gr(n, p):
       Space of p-dimensional subspaces of ℝ^n.
       Used for tracking low-rank subspaces in weight matrices.
       Projection: Π_X(Z) = Z - X(X^T Z)

Supported Manifolds:
    - Stiefel(n, p):   Orthogonality-constrained weight matrices
    - Sphere(n):       Unit-norm embedding vectors (S^{n-1})
    - Grassmann(n, p): Low-rank subspace tracking
    - Euclidean(n):    Standard flat space (fallback)

Reference: Absil, Mahony, Sepulchre — "Optimization Algorithms on Matrix Manifolds" (2008)
"""
from __future__ import annotations

import math
from enum import Enum
from typing import Dict, List, Optional, Tuple

import numpy as np


class ManifoldType(Enum):
    EUCLIDEAN = "euclidean"
    SPHERE = "sphere"
    STIEFEL = "stiefel"
    GRASSMANN = "grassmann"


# ─────────────────────────────────────────────────────────────────────
# Manifold Operations
# ─────────────────────────────────────────────────────────────────────

def _stiefel_project_tangent(X: np.ndarray, Z: np.ndarray) -> np.ndarray:
    """
    Project Z onto tangent space T_X St(n, p).

    For Stiefel manifold St(n, p) = {X : X^T X = I_p}:
        Π_X(Z) = Z - X · sym(X^T Z)
    where sym(A) = (A + A^T) / 2.
    """
    XtZ = X.T @ Z
    sym = 0.5 * (XtZ + XtZ.T)
    return Z - X @ sym


def _stiefel_retract(X: np.ndarray, Z: np.ndarray) -> np.ndarray:
    """
    Cayley retraction on Stiefel manifold.

    R_X(Z) = (I - Z/2)^{-1} (I + Z/2) X
    This is a first-order approximation to the exponential map,
    cheaper than QR or SVD-based retractions.

    Actually for numerical stability, use QR:
        X_new = qr(X + Z).Q
    """
    Y = X + Z
    Q, R = np.linalg.qr(Y)
    # Fix sign ambiguity: ensure diagonal of R is positive
    signs = np.sign(np.diag(R))
    signs[signs == 0] = 1.0
    Q *= signs[np.newaxis, :]
    return Q


def _sphere_project_tangent(x: np.ndarray, z: np.ndarray) -> np.ndarray:
    """
    Project z onto tangent space T_x S^{n-1}.

    Π_x(z) = z - (x · z) x
    """
    return z - np.dot(x, z) * x


def _sphere_retract(x: np.ndarray, z: np.ndarray) -> np.ndarray:
    """
    Retraction on the sphere: normalise after step.

    R_x(z) = (x + z) / ‖x + z‖

    This is the cheapest retraction and first-order accurate.
    """
    y = x + z
    norm = np.linalg.norm(y)
    if norm < 1e-12:
        return x
    return y / norm


def _grassmann_project_tangent(X: np.ndarray, Z: np.ndarray) -> np.ndarray:
    """
    Project Z onto tangent space T_X Gr(n, p).

    Π_X(Z) = (I - XX^T) Z = Z - X(X^T Z)
    """
    return Z - X @ (X.T @ Z)


def _grassmann_retract(X: np.ndarray, Z: np.ndarray) -> np.ndarray:
    """
    QR-based retraction on Grassmann manifold.
    """
    Y = X + Z
    Q, _ = np.linalg.qr(Y)
    return Q


def _parallel_transport(x: np.ndarray, y: np.ndarray,
                        v: np.ndarray) -> np.ndarray:
    """
    Vector transport by projection (generic).

    Γ_{x→y}(v) = Π_{T_y}(v)

    For the sphere, this is:
        v' = v - (y · v) y

    For general manifolds, subtract the normal component at y.
    """
    delta = y - x
    norm_sq = np.dot(delta, delta) if delta.ndim == 1 else np.sum(delta ** 2)
    if norm_sq < 1e-24:
        return v.copy()
    # Householder-type transport
    return v - (np.dot(delta, v) / norm_sq) * delta if v.ndim == 1 else \
        v - (delta @ (delta.T @ v)) / norm_sq


# ─────────────────────────────────────────────────────────────────────
# Riemannian Optimizer
# ─────────────────────────────────────────────────────────────────────

class RiemannianSGD:
    """
    Stochastic Gradient Descent on Riemannian manifolds.

    Supports momentum with parallel transport:
        v_t = β · Γ(v_{t-1}) + ∇_M f(x_t)
        x_{t+1} = Ret_x(-η · v_t)

    Parameters
    ----------
    lr : float
        Learning rate η.
    momentum : float
        Momentum coefficient β ∈ [0, 1).
    manifold : ManifoldType
        Which manifold the parameters live on.
    """

    def __init__(self, lr: float = 0.01, momentum: float = 0.9,
                 manifold: ManifoldType = ManifoldType.STIEFEL):
        self.lr = lr
        self.momentum = momentum
        self.manifold = manifold
        self._velocity: Optional[np.ndarray] = None
        self._prev_point: Optional[np.ndarray] = None

        # Statistics
        self.steps = 0
        self.total_grad_norm = 0.0

    def step(self, X: np.ndarray, grad: np.ndarray) -> np.ndarray:
        """
        One Riemannian gradient step.

        Parameters
        ----------
        X : np.ndarray
            Current point on the manifold.
        grad : np.ndarray
            Euclidean gradient ∇f(X) (will be projected to tangent space).

        Returns
        -------
        X_new : np.ndarray
            Updated point on the manifold.
        """
        # 1. Project gradient to tangent space: ∇_M f = Π_{T_X}(∇f)
        rgrad = self._project_tangent(X, grad)

        # Track statistics
        self.total_grad_norm += float(np.linalg.norm(rgrad))
        self.steps += 1

        # 2. Momentum with parallel transport
        if self._velocity is not None and self._prev_point is not None:
            transported_v = _parallel_transport(self._prev_point, X, self._velocity)
            # Re-project to ensure we stay in tangent space
            transported_v = self._project_tangent(X, transported_v)
            velocity = self.momentum * transported_v + rgrad
        else:
            velocity = rgrad.copy()

        self._velocity = velocity.copy()
        self._prev_point = X.copy()

        # 3. Retraction: x_{t+1} = Ret_X(-η · v)
        X_new = self._retract(X, -self.lr * velocity)

        return X_new

    def _project_tangent(self, X: np.ndarray, Z: np.ndarray) -> np.ndarray:
        if self.manifold == ManifoldType.STIEFEL:
            return _stiefel_project_tangent(X, Z)
        elif self.manifold == ManifoldType.SPHERE:
            return _sphere_project_tangent(X, Z)
        elif self.manifold == ManifoldType.GRASSMANN:
            return _grassmann_project_tangent(X, Z)
        else:
            return Z  # Euclidean: no projection needed

    def _retract(self, X: np.ndarray, Z: np.ndarray) -> np.ndarray:
        if self.manifold == ManifoldType.STIEFEL:
            return _stiefel_retract(X, Z)
        elif self.manifold == ManifoldType.SPHERE:
            return _sphere_retract(X, Z)
        elif self.manifold == ManifoldType.GRASSMANN:
            return _grassmann_retract(X, Z)
        else:
            return X + Z  # Euclidean

    def avg_grad_norm(self) -> float:
        return self.total_grad_norm / max(self.steps, 1)


# ─────────────────────────────────────────────────────────────────────
# Riemannian Adam (RAdam)
# ─────────────────────────────────────────────────────────────────────

class RiemannianAdam:
    """
    Adam optimizer on Riemannian manifolds.

    Extends Adam with:
        1. Riemannian gradient (tangent projection)
        2. Exponential moving averages transported along geodesics
        3. Bias-corrected moment estimates on the manifold

    Update rule:
        g_t = Π_{T_x}(∇f(x_t))
        m_t = β₁ · Γ(m_{t-1}) + (1 - β₁) · g_t
        v_t = β₂ · v_{t-1} + (1 - β₂) · ‖g_t‖²   (scalar, no transport needed)
        m̂_t = m_t / (1 - β₁^t)
        v̂_t = v_t / (1 - β₂^t)
        x_{t+1} = Ret_x(-η · m̂_t / (√v̂_t + ε))
    """

    def __init__(self, lr: float = 0.001, beta1: float = 0.9,
                 beta2: float = 0.999, eps: float = 1e-8,
                 manifold: ManifoldType = ManifoldType.STIEFEL):
        self.lr = lr
        self.beta1 = beta1
        self.beta2 = beta2
        self.eps = eps
        self.manifold = manifold

        self._m: Optional[np.ndarray] = None  # First moment
        self._v: float = 0.0                  # Second moment (scalar)
        self._prev_point: Optional[np.ndarray] = None
        self._t: int = 0

    def step(self, X: np.ndarray, grad: np.ndarray) -> np.ndarray:
        """One Riemannian Adam step."""
        self._t += 1

        # Project to tangent space
        rgrad = self._project_tangent(X, grad)
        grad_norm_sq = float(np.sum(rgrad ** 2))

        # Update moments
        if self._m is not None and self._prev_point is not None:
            transported_m = _parallel_transport(self._prev_point, X, self._m)
            transported_m = self._project_tangent(X, transported_m)
            self._m = self.beta1 * transported_m + (1 - self.beta1) * rgrad
        else:
            self._m = (1 - self.beta1) * rgrad

        self._v = self.beta2 * self._v + (1 - self.beta2) * grad_norm_sq

        # Bias correction
        m_hat = self._m / (1 - self.beta1 ** self._t)
        v_hat = self._v / (1 - self.beta2 ** self._t)

        # Step
        step_size = self.lr / (math.sqrt(v_hat) + self.eps)
        X_new = self._retract(X, -step_size * m_hat)

        self._prev_point = X.copy()
        return X_new

    def _project_tangent(self, X, Z):
        if self.manifold == ManifoldType.STIEFEL:
            return _stiefel_project_tangent(X, Z)
        elif self.manifold == ManifoldType.SPHERE:
            return _sphere_project_tangent(X, Z)
        elif self.manifold == ManifoldType.GRASSMANN:
            return _grassmann_project_tangent(X, Z)
        return Z

    def _retract(self, X, Z):
        if self.manifold == ManifoldType.STIEFEL:
            return _stiefel_retract(X, Z)
        elif self.manifold == ManifoldType.SPHERE:
            return _sphere_retract(X, Z)
        elif self.manifold == ManifoldType.GRASSMANN:
            return _grassmann_retract(X, Z)
        return X + Z

File: kernel/epsilon/epsilon/crates/aether-core/src/riemannian_optimizer.rs
// Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
// SPDX-License-Identifier: Epsilon-Hollow

//! Riemannian optimizer primitives migrated from legacy host code.
//!
//! This module keeps the production theorem/runtime path in no-std Rust:
//! fixed-size arrays, no heap allocation, and explicit manifold projection /
//! retraction formulas. Distributed Stiefel synchronization remains in
//! [`crate::parallel_riemannian`]; this file owns local optimizer state.

use libm::{pow, sqrt};

const EPS: f64 = 1e-12;

/// Supported local optimizer manifolds.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ManifoldType {
    /// Plain Euclidean space; no tangent projection or curved retraction.
    Euclidean,
    /// Unit sphere with tangent projection `z - <x,z>x`.
    Sphere,
    /// Stiefel manifold. Matrix helper APIs preserve the theorem formulas.
    Stiefel,
    /// Grassmann manifold. Matrix helper APIs preserve the theorem formulas.
    Grassmann,
}

/// Project `z` onto the tangent space of the unit sphere at `x`.
pub fn sphere_project_tangent<const D: usize>(x: &[f64; D], z: &[f64; D]) -> [f64; D] {
    let dot = dot_vector(x, z);
    let mut projected = [0.0_f64; D];
    for idx in 0..D {
        projected[idx] = z[idx] - dot * x[idx];
    }
    projected
}

/// Retract a tangent vector on the sphere by normalizing `x + z`.
pub fn sphere_retract<const D: usize>(x: &[f64; D], z: &[f64; D]) -> [f64; D] {
    let mut y = [0.0_f64; D];
    for idx in 0..D {
        y[idx] = x[idx] + z[idx];
    }

    let norm = vector_norm(&y);
    if norm < EPS {
        return *x;
    }

    for value in &mut y {
        *value /= norm;
    }
    y
}

/// Transport a sphere tangent vector by projecting it onto `T_y S^(D-1)`.
///
/// The legacy documentation describes this transport as:
///
/// ```text
/// Gamma_{x -> y}(v) = Pi_{T_y}(v) = v - <y,v>y
/// ```
pub fn parallel_transport_sphere<const D: usize>(
    _x: &[f64; D],
    y: &[f64; D],
    v: &[f64; D],
) -> [f64; D] {
    sphere_project_tangent(y, v)
}

/// Project `z` onto `T_x St(n,p)` with `Pi_X(Z) = Z - X sym(X^T Z)`.
pub fn stiefel_project_tangent<const N: usize, const P: usize>(
    x: &[[f64; P]; N],
    z: &[[f64; P]; N],
) -> [[f64; P]; N] {
    let mut xtz = [[0.0_f64; P]; P];
    for row in 0..N {
        for col_x in 0..P {
            for col_z in 0..P {
                xtz[col_x][col_z] += x[row][col_x] * z[row][col_z];
            }
        }
    }

    let mut sym = [[0.0_f64; P]; P];
    for row in 0..P {
        for col in 0..P {
            sym[row][col] = 0.5 * (xtz[row][col] + xtz[col][row]);
        }
    }

    let mut projected = *z;
    for row in 0..N {
        for col in 0..P {
            let mut correction = 0.0;
            for mid in 0..P {
                correction += x[row][mid] * sym[mid][col];
            }
            projected[row][col] -= correction;
        }
    }
    projected
}

/// QR-style Stiefel retraction using modified Gram-Schmidt on `x + z`.
pub fn stiefel_retract<const N: usize, const P: usize>(
    x: &[[f64; P]; N],
    z: &[[f64; P]; N],
) -> [[f64; P]; N] {
    let mut y = *x;
    for row in 0..N {
        for col in 0..P {
            y[row][col] += z[row][col];
        }
    }
    orthonormalize_columns(y)
}

/// Project `z` onto `T_x Gr(n,p)` with `Pi_X(Z) = Z - X(X^T Z)`.
pub fn grassmann_project_tangent<const N: usize, const P: usize>(
    x: &[[f64; P]; N],
    z: &[[f64; P]; N],
) -> [[f64; P]; N] {
    let mut xtz = [[0.0_f64; P]; P];
    for row in 0..N {
        for col_x in 0..P {
            for col_z in 0..P {
                xtz[col_x][col_z] += x[row][col_x] * z[row][col_z];
            }
        }
    }

    let mut projected = *z;
    for row in 0..N {
        for col in 0..P {
            let mut correction = 0.0;
            for mid in 0..P {
                correction += x[row][mid] * xtz[mid][col];
            }
            projected[row][col] -= correction;
        }
    }
    projected
}

/// QR-style Grassmann retraction using modified Gram-Schmidt on `x + z`.
pub fn grassmann_retract<const N: usize, const P: usize>(
    x: &[[f64; P]; N],
    z: &[[f64; P]; N],
) -> [[f64; P]; N] {
    stiefel_retract(x, z)
}

/// Fixed-size Riemannian SGD with optional momentum.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct RiemannianSgd<const D: usize> {
    lr: f64,
    momentum: f64,
    manifold: ManifoldType,
    velocity: [f64; D],
    previous_point: Option<[f64; D]>,
    steps: u32,
    total_grad_norm: f64,
}

impl<const D: usize> RiemannianSgd<D> {
    /// Create an SGD optimizer for fixed-size vector parameters.
    pub const fn new(lr: f64, momentum: f64, manifold: ManifoldType) -> Self {
        Self {
            lr,
            momentum,
            manifold,
            velocity: [0.0; D],
            previous_point: None,
            steps: 0,
            total_grad_norm: 0.0,
        }
    }

    /// Apply one Riemannian gradient step to vector parameters.
    pub fn step_vector(&mut self, x: [f64; D], grad: [f64; D]) -> [f64; D] {
        let rgrad = project_vector(self.manifold, &x, &grad);
        self.total_grad_norm += vector_norm(&rgrad);
        self.steps = self.steps.saturating_add(1);

        let transported = self
            .previous_point
            .map(|prev| transport_vector(self.manifold, &prev, &x, &self.velocity))
            .unwrap_or([0.0; D]);

        let mut velocity = [0.0_f64; D];
        for idx in 0..D {
            velocity[idx] = if self.previous_point.is_some() {
                self.momentum * transported[idx] + rgrad[idx]
            } else {
                rgrad[idx]
            };
        }

        self.velocity = velocity;
        self.previous_point = Some(x);

        let mut step = [0.0_f64; D];
        for idx in 0..D {
            step[idx] = -self.lr * velocity[idx];
        }
        retract_vector(self.manifold, &x, &step)
    }

    /// Number of optimizer steps applied.
    pub const fn steps(&self) -> u32 {
        self.steps
    }

    /// Average norm of projected gradients observed so far.
    pub fn avg_grad_norm(&self) -> f64 {
        if self.steps == 0 {
            0.0
        } else {
            self.total_grad_norm / f64::from(self.steps)
        }
    }
}

impl<const D: usize> Default for RiemannianSgd<D> {
    fn default() -> Self {
        Self::new(0.01, 0.9, ManifoldType::Stiefel)
    }
}

/// Fixed-size Riemannian Adam with scalar second moment.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct RiemannianAdam<const D: usize> {
    lr: f64,
    beta1: f64,
    beta2: f64,
    eps: f64,
    manifold: ManifoldType,
    first_moment: [f64; D],
    second_moment: f64,
    previous_point: Option<[f64; D]>,
    steps: u32,
}

impl<const D: usize> RiemannianAdam<D> {
    /// Create an Adam optimizer for fixed-size vector parameters.
    pub const fn new(lr: f64, beta1: f64, beta2: f64, eps: f64, manifold: ManifoldType) -> Self {
        Self {
            lr,
            beta1,
            beta2,
            eps,
            manifold,
            first_moment: [0.0; D],
            second_moment: 0.0,
            previous_point: None,
            steps: 0,
        }
    }

    /// Apply one Riemannian Adam step to vector parameters.
    pub fn step_vector(&mut self, x: [f64; D], grad: [f64; D]) -> [f64; D] {
        self.steps = self.steps.saturating_add(1);
        let rgrad = project_vector(self.manifold, &x, &grad);
        let transported = self
            .previous_point
            .map(|prev| transport_vector(self.manifold, &prev, &x, &self.first_moment))
            .unwrap_or([0.0; D]);

        for idx in 0..D {
            self.first_moment[idx] = if self.previous_point.is_some() {
                self.beta1 * transported[idx] + (1.0 - self.beta1) * rgrad[idx]
            } else {
                (1.0 - self.beta1) * rgrad[idx]
            };
        }

        self.second_moment =
            self.beta2 * self.second_moment + (1.0 - self.beta2) * squared_norm(&rgrad);

        let bias1 = 1.0 - pow(self.beta1, f64::from(self.steps));
        let bias2 = 1.0 - pow(self.beta2, f64::from(self.steps));
        let inv_bias1 = if bias1.abs() < EPS { 1.0 } else { 1.0 / bias1 };
        let v_hat = if bias2.abs() < EPS {
            self.second_moment
        } else {
            self.second_moment / bias2
        };
        let scale = self.lr / (sqrt(v_hat) + self.eps);

        let mut step = [0.0_f64; D];
        for idx in 0..D {
            step[idx] = -scale * self.first_moment[idx] * inv_bias1;
        }

        self.previous_point = Some(x);
        retract_vector(self.manifold, &x, &step)
    }

    /// Number of optimizer steps applied.
    pub const fn steps(&self) -> u32 {
        self.steps
    }
}

impl<const D: usize> Default for RiemannianAdam<D> {
    fn default() -> Self {
        Self::new(0.001, 0.9, 0.999, 1e-8, ManifoldType::Stiefel)
    }
}

fn project_vector<const D: usize>(
    manifold: ManifoldType,
    x: &[f64; D],
    grad: &[f64; D],
) -> [f64; D] {
    match manifold {
        ManifoldType::Sphere => sphere_project_tangent(x, grad),
        ManifoldType::Euclidean | ManifoldType::Stiefel | ManifoldType::Grassmann => *grad,
    }
}

fn retract_vector<const D: usize>(
    manifold: ManifoldType,
    x: &[f64; D],
    step: &[f64; D],
) -> [f64; D] {
    match manifold {
        ManifoldType::Sphere => sphere_retract(x, step),
        ManifoldType::Euclidean | ManifoldType::Stiefel | ManifoldType::Grassmann => {
            let mut y = [0.0_f64; D];
            for idx in 0..D {
                y[idx] = x[idx] + step[idx];
            }
            y
        }
    }
}

fn transport_vector<const D: usize>(
    manifold: ManifoldType,
    prev: &[f64; D],
    x: &[f64; D],
    v: &[f64; D],
) -> [f64; D] {
    match manifold {
        ManifoldType::Sphere => parallel_transport_sphere(prev, x, v),
        ManifoldType::Euclidean | ManifoldType::Stiefel | ManifoldType::Grassmann => *v,
    }
}

fn orthonormalize_columns<const N: usize, const P: usize>(mut q: [[f64; P]; N]) -> [[f64; P]; N] {
    for col in 0..P {
        for prev in 0..col {
            let mut projection = 0.0;
            for row in 0..N {
                projection += q[row][prev] * q[row][col];
            }
            for row in 0..N {
                q[row][col] -= projection * q[row][prev];
            }
        }

        let mut norm_sq = 0.0;
        for row in 0..N {
            norm_sq += q[row][col] * q[row][col];
        }
        let norm = sqrt(norm_sq);

        if norm < EPS {
            for row in 0..N {
                q[row][col] = 0.0;
            }
            if col < N {
                q[col][col] = 1.0;
            }
        } else {
            for row in 0..N {
                q[row][col] /= norm;
            }
        }
    }
    q
}

fn dot_vector<const D: usize>(a: &[f64; D], b: &[f64; D]) -> f64 {
    let mut sum = 0.0;
    for idx in 0..D {
        sum += a[idx] * b[idx];
    }
    sum
}

fn squared_norm<const D: usize>(v: &[f64; D]) -> f64 {
    dot_vector(v, v)
}

fn vector_norm<const D: usize>(v: &[f64; D]) -> f64 {
    sqrt(squared_norm(v))
}
